### DISCUSSION TITLE
Steps/Question about moving FW update to fwupd

### DISCUSSION THREAD

**Discussion by lwang47-amr** [fwupd/fwupd#7212] [General]
This is a continuous discussion from https://gitlab.com/kernel-firmware/linux-firmware/-/merge_requests/207

I'm looking into a way to build a simple bare-minimal prototype (seleton) for Intel CV FW device update solution to show case the solution.

To start with, I'll 

1. expose an API from device driver, so that fwupdmgr can write FW binary to it via /sys fs
2. prepare the plug-in in fwupd, which can parse & handle the binary file from user space.

To build and test a bare-minimal environment, I'd like to do it without LFVS, by running 
"fwupdtool --verbose --plugins intel-cvs install-blob /tmp/cvs-firmware.bin $CVS_DEVICE_ID", is this correct?
 
Also, two other questions,

1. How do i get my CV's DEVICE_ID? This sounds like need to LVFS meta file, which I don't and won't have.
2. Any way I can download a device FW binary into a local dir, say "/tmp"? I tried "fwupdmgr download file:///tmp" but "fwupdmgr --reinstall" didn't download the file. This is, of course, to another working device (not CV)

**Comment by superm1** [comment]
If you haven't already, you should read over the plugin tutorial:
https://fwupd.github.io/libfwupdplugin/tutorial.html

> expose an API from device driver, so that fwupdmgr can write FW binary to it via /sys fs

Sounds reasonable.  You can add a dummy sysfs to your driver to POC how it would work and allow you to do the fwupd plugin in parallel.

> prepare the plug-in in fwupd, which can parse & handle the binary file from user space.

As I mentioned in the other thread I think this could mostly model off of how the "amd-gpu" plugin works.   We have a quirk file in the plugin that will make sure that it runs anytime a device is bound with the `amdgpu` kernel driver: https://github.com/fwupd/fwupd/blob/main/plugins/amd-gpu/amd-gpu.quirk

The constructed callback for the plugin will make sure that matching udev events come to the plugin.
https://github.com/fwupd/fwupd/blob/main/plugins/amd-gpu/fu-amd-gpu-plugin.c#L34

> How do i get my CV's DEVICE_ID? This sounds like need to LVFS meta file, which I don't and won't have.

The instance ID is built off of whatever you use to identify the hardware.  If you have unique PCI IDs for each stream of CV firmware you can use those.  For example on my system I have some devices that make the instance IDs and GUIDs this way:
```
│ │     GUIDs:            9eb6a793-7f97-5fb5-b49c-139d2ce3ee46 ← PCI\VEN_1022&DEV_15C7
│ │                       2e4eb727-a7c8-5eba-80cd-553b8d796d80 ← PCI\VEN_1022&DEV_15C7&SUBSYS_F1110006
```

If you need to add another attribute into the instance ID you can based of whatever you want.  The plugin I linked showed using libdrm to get it:
```
│ │ │   GUID:             6c23af4f-d6cd-5ffc-a502-0b85e472e7cb ← AMD\113-PHXGENERIC-001
```
It can just as well be a string from a sysfs attribute or any other API.

> Any way I can download a device FW binary into a local dir, say "/tmp"? I tried "fwupdmgr download file:///tmp" but "fwupdmgr --reinstall" didn't download the file. This is, of course, to another working device (not CV)

This is putting the cart before the horse, you should work out how the plugin should work, and then you can use `fwupdtool install-blob foo.bin` to try to install a fake binary.  I suggest modifying the kernel driver to print the header of the binary to `dmesg` or something similar to verify the data made it to the kernel correctly.
